File: frontEnd/src/app/services/hoivien.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { HoiVien } from '../models/hoi-vien';

@Injectable({
  providedIn: 'root'
})
export class HoivienService {
  private apiUrl = 'http://localhost:8080/hoiVien'; 
  constructor(private http: HttpClient) { }
  hoiVien: any = {
    maHV: "",
    email: "",
    soDienThoai: "",
    hoTen: "",
    diaChi: "",
    thoiGianDangKy: new Date(),
    tinhTrang: true,
    hinhAnhHV: "",
    tienNap:0,
    taiKhoanHV: {
        userName: "",
        password: "",
        authorities: []
    }
};
  searchByHoTen(name: string): Observable<HoiVien[]> {
    return this.http.get<any[]>(`${this.apiUrl}/search?name=${name}`);
  }
  findById(maHV: string): Observable<HoiVien> {
    return this.http.get<HoiVien>(`${this.apiUrl}/${maHV}`);
  }
  findByUsername(username: string): Observable<any> {
    return this.http.get<HoiVien>(`${this.apiUrl+'/username'}/${username}`);
  }
  updateHoiVien(hoivien: any,  file: File | null): Observable<any> {  
    const formData = new FormData();
    formData.append('hoivien', JSON.stringify(hoivien));
        // Chỉ thêm file vào formData nếu nó không phải là null
        if (file) {
          formData.append('file', file);
      }
    return this.http.put(this.apiUrl, formData);
  }
  capnhatHoiVien(userName:string,hoiVien: any, file: File | null): Observable<any> {
    const formData = new FormData();
    formData.append('hoivien', JSON.stringify(hoiVien));
    formData.append('userName', userName);
    // Chỉ thêm file vào formData nếu nó không phải là null
    if (file) {
        formData.append('file', file);
    }
    return this.http.put(this.apiUrl+'/update', formData);
  }
  getHoiVien(): Observable<any[]> {  
    return this.http.get<any[]>(this.apiUrl);
  }
  addHoiVien(hoivien: any,  file: File | null): Observable<any> {  
    const formData = new FormData();
    formData.append('hoivien', JSON.stringify(hoivien));
        // Chỉ thêm file vào formData nếu nó không phải là null
        if (file) {
          formData.append('file', file);
      }
    return this.http.post(this.apiUrl, formData);
  }

  deleteHoiVien(mahoivien: string): Observable<any> {
    return this.http.delete(`${this.apiUrl}/${mahoivien}`);
  }
      // Hàm gọi API để tìm kiếm với phân trang
      searchHoiViens(keyword: string, page: number, size: number): Observable<any> {
        let params = new HttpParams()
          .set('keyword', keyword || '')  // Nếu không có keyword, gửi chuỗi rỗng
          .set('page', page.toString())
          .set('size', size.toString());
    
        return this.http.get<any>(this.apiUrl+"/searchPage", { params });
      }

}

File: BackEnd/BookLand/src/main/java/com/bookland/service/hoiVienService.java
package com.bookland.service;

import java.util.List;
import java.util.Optional;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import com.bookland.entity.HoiVien;

@Service 
public interface hoiVienService {

	List<HoiVien> findHoiVienByName(String name);

	 Optional<HoiVien> findById(String maHV);

//	 Optional<HoiVien> findById(String maHV);
		 Optional<HoiVien> findByUsername(String username); // tìm hội viên dựa vào username	    
//		 Optional<HoiVien> findById(String maHV);		
		 public HoiVien updateHV(String userName,HoiVien hoivien,MultipartFile file);
	     public List <HoiVien> finAll();
		 public HoiVien create(HoiVien hoivien,MultipartFile file);
		 public void delete(String maHoiVien); 
		 public HoiVien  update(HoiVien hoiVien,MultipartFile file); 
		 Page<HoiVien> searchHoiViens(String keyword, Pageable pageable);
	

}
